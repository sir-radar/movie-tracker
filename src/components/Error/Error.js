import React from 'react'

const Error = () => {
  return (
  <section data-testid="error" className='loading mt-5 mx-auto text-center'>
    <i className="fa fa-times fa-5x text-danger"></i>
    <h1 data-testid="error-header">Opps</h1>
    <h3 data-testid="error-body">Something went wrong. Please check your internet and try again.</h3>
  </section>
  )
}

export default Error