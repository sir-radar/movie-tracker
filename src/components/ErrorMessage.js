import React,{useEffect} from 'react'

const ErrorMessage = ({resetStatus}) => {
  //resets error status to empty
  useEffect(() => {
    setTimeout(()=>{
      resetStatus()
    }, 2000)
  },[resetStatus]);

  return (
    <div className="w-25 bg-danger text-white text-center message">
      Something went wrong
    </div>
  )
}

export default ErrorMessage