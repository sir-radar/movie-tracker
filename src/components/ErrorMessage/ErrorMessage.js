import React,{useEffect} from 'react'

const ErrorMessage = ({resetStatus}) => {
  //resets error status to empty
  useEffect(() => {
    setTimeout(()=>{
      resetStatus()
    }, 2000)
  },[resetStatus]);

  return (
    <div data-testid="error-message" className="w-auto bg-danger p-1 text-white text-center message">
      Something went wrong
    </div>
  )
}

export default ErrorMessage