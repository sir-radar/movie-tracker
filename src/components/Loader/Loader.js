import React from 'react';

const Loader = () => {
  return (
    <section className='loading mt-5 mx-auto'>
      <i className="fa fa-spinner fa-5x fa-spin"></i>
    </section>
  );
}

export default Loader;
