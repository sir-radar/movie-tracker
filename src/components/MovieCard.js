import React from 'react';

const MovieCard = ({id, title, image, overview,addToFavorite, removeFavorite, favoriteIDs}) => {

  const truncate = (str) => str.length > 100 ? str.substring(0, 100) + "..." : str;

  const makeFavourite = (e) =>{
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": true
    };
    e.preventDefault()
    addToFavorite(data)
  }

  const removeFavourite = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": false
    };
    removeFavorite(data)
  }

  return (
    <div className="card card-height p-0 col-12 col-md-5 m-2">
      <img src={image ? `https://image.tmdb.org/t/p/w200/${image}` : 'no-image.png'} className="card-img-top image-size" alt="..."/>
      <div className="d-flex flex-column justify-content-end card-body">
        <h5 className="card-title">{title}</h5>
        <div>
          <p className="card-text">{truncate(overview)}</p>
        </div>
        <div className="align-self-center actions">
          <a href="#" className="btn btn-primary">Go somewhere</a>
          <div className="d-flex">
            {
              favoriteIDs.includes(id) ? <button onClick={removeFavourite} className="btn mr-2"><i className="fa fa-star text-favourite"></i></button>
                : <button onClick={makeFavourite} className="btn mr-2"><i className="fa fa-star"></i></button>
            }
            
            <button className="btn">Watch later</button>
          </div>
        </div>
      </div>
    </div>
  );
}

export default MovieCard;
