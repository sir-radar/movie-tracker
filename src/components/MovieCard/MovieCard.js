import React, {useEffect, useState} from 'react';
import { connect } from 'react-redux';
import {Link} from 'react-router-dom';
import { getMovieState } from '../../store/actions/movieStateAction';

const MovieCard = (props) => {

  const { id,
          title, 
          image, 
          overview, 
          addToFavourite, 
          removeFavourite, 
          removeFromWatchList,
          addToWatchList,
          getMovieState
        } = props

  const [movieState, setMovieState] = useState({})
//  let movieState = {};

  useEffect(() => {
    //prevents memory leaks and prevents settings state for unrendered component
    let mounted = true;

    async function movieStateRequest(){
      const response = await getMovieState(id)
      if(mounted){
        setMovieState(response.data)
      }
    }
    movieStateRequest()

    return () => mounted = false;
    
  });

  //Limits text to the length
  const truncate = (str, length) => str.length > length ? str.substring(0, length) + "..." : str;

  const makeFavouriteHandler = (e) =>{
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": true
    };
    e.preventDefault()
    addToFavourite(data)
  }

  const removeFromFavouritesHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "favorite": false
    };
    removeFavourite(data)
  }

  const addToWatchListHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "watchlist": true
    };
    addToWatchList(data)
  }

  const removeFromWatchListHandler = () => {
    const data = {
      "media_type": "movie",
      "media_id": id,
      "watchlist": false
    };
    removeFromWatchList(data)
  }

  return (
    <div data-testid="movie-card" className="col-md-3 mt-2">
      <div className="card card-height p-0 col-12">
        <img src={image ? `https://image.tmdb.org/t/p/w500/${image}` : '/no-image.png'} className="card-img-top image-size mb-2" alt="..."/>
        <div className="d-flex flex-column justify-content-between card-body p-2">
          <h5 data-testid="movie-title" className="card-title">{truncate(title, 30)}</h5>
          <div>
            <p data-testid="movie-overview" className="card-text">{truncate(overview, 70)}</p>
          </div>
          <div className="align-self-center actions text-center">
            <Link data-testid="movie-details-link" to={`movie/${id}`}>See Details</Link>
            <div className="d-flex mt-2">
              {
                movieState.favorite ? <button data-testid="active-favourite" onClick={removeFromFavouritesHandler} className="btn mr-2"><i className="fa fa-star text-favourite"></i></button>
                  : <button data-testid="inactive-favourite" onClick={makeFavouriteHandler} className="btn mr-2"><i className="fa fa-star"></i></button>
              }
              {
                movieState.watchlist ? <button data-testid="active-watchlist" onClick={removeFromWatchListHandler} className="btn rm-btn-bg">Remove</button>
                  : <button data-testid="inactive-watchlist" onClick={addToWatchListHandler} className="btn">Watch later</button>
              }
            </div>
          </div>
        </div>
      </div>
    </div>
  );
}

const mapDispatchToProps = (dispatch) => (
  {
    getMovieState: (movieId) => dispatch(getMovieState(movieId)),
  }
)

export default connect(null, mapDispatchToProps) (MovieCard);
