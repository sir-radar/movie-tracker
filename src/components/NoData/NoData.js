import React from 'react';

const NoData = () => {
  return (
    <section data-testid="nodata-component" className='no-data mt-5 mx-auto'>
      <h2 data-testid="nodata">No data to display</h2>
    </section>
  );
}

export default NoData;
