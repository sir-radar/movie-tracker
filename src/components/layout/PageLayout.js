import React from 'react'
import Nav from '../Nav/Nav';

const PageLayout = (props) => {
  return (
    <>
      { props.showNav?<Nav/> : null}
      <main className="mt-4 p-3 col-10 col-md-10 offset-1 offset-md-1">
          {props.searchbox}
        <div className="row">
          {props.pageContent}
          {props.loader}
          {props.error}
          {props.nodata}
          {props.pagination}
          {props.errorMessage}
        </div>
      </main>
    </>
  )
}

export default PageLayout