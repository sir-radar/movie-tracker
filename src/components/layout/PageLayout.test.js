import React from 'react';
import { MemoryRouter } from "react-router-dom";
import { render, cleanup } from '@testing-library/react';
import PageLayout from './PageLayout';
import SearchBox from '../SearchBox/SearchBox';
import MovieCard from '../MovieCard/MovieCard';
import Loader from '../Loader/Loader';
import Error from '../Error/Error';
import NoData from '../NoData/NoData';
import Pagination from '../Pagination/Pagination';
import ErrorMessage from '../ErrorMessage/ErrorMessage';
import "@testing-library/jest-dom/extend-expect";
import { Provider } from 'react-redux';
import renderer from "react-test-renderer";
import thunk from "redux-thunk";
import configureStore from "redux-mock-store";

const buildStore = configureStore([thunk]);


afterEach(cleanup)

const renderComponent = (props={}) => {
  const states = {
    search:{},
    favourites:{},
    status:{},
    watchlists:{},
    auth:{
      session_id: "kkekdllldldll"
    }
  }
  const store = buildStore({...states});
  const div = document.createElement("div");
  return render(<Provider store={store}><MemoryRouter><PageLayout {...props}/></MemoryRouter></Provider>, div)
}

describe('PageLayout Component', () => {

  it("renders without crashing", () => {
    renderComponent()
  })

  it("renders without any content", () => {
    const {queryByTestId} = renderComponent()
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("render nav component when showNav props is passed into it ", () => {
    const {queryByTestId} = renderComponent({showNav: true})
    expect(queryByTestId("nav")).not.toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders seach box component when passed as props", () => {
    const { queryByTestId } = renderComponent({searchbox: <SearchBox/>})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).not.toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders MoviesCard component when passed as props", () => {
    const movieCarPpropsData = {
      id: 1,
      title: "Movie title", 
      image: "", 
      overview: "Movie overview", 
      addToFavourite: jest.fn(), 
      removeFavourite: jest.fn(), 
      removeFromWatchList: jest.fn(),
      addToWatchList: jest.fn(),
    };
    const { queryByTestId } = renderComponent({pageContent: <MovieCard {...movieCarPpropsData}/>})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).not.toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders Loader component when passed as props", () => {
    const { queryByTestId } = renderComponent({loader: <Loader />})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("loader")).not.toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders Error component when passed as props", () => {
    const { queryByTestId } = renderComponent({error: <Error />})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).not.toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders NoData component when passed as props", () => {
    const { queryByTestId } = renderComponent({nodata: <NoData />})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).not.toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders Pagination component when passed as props", () => {
    const { queryByTestId } = renderComponent({pagination: <Pagination />})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).not.toBeNull()
    expect(queryByTestId("error-message")).toBeNull()
  })

  it("renders ErrorMessage component when passed as props", () => {
    const { queryByTestId } = renderComponent({errorMessage: <ErrorMessage />})
    expect(queryByTestId("nav")).toBeNull()
    expect(queryByTestId("search-box")).toBeNull()
    expect(queryByTestId("movie-card")).toBeNull()
    expect(queryByTestId("loader")).toBeNull()
    expect(queryByTestId("error")).toBeNull()
    expect(queryByTestId("nodata-component")).toBeNull()
    expect(queryByTestId("pagination")).toBeNull()
    expect(queryByTestId("error-message")).not.toBeNull()
  })


  it("matches snapshot", () => {
    const tree = renderer.create(<MemoryRouter><PageLayout/></MemoryRouter>).toJSON();
    expect(tree).toMatchSnapshot();
  })

})