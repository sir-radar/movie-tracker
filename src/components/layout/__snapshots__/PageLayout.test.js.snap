// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLayout Component matches snapshot 1`] = `
<main
  className="mt-4 p-3 col-10 col-md-10 offset-1 offset-md-1"
>
  <div
    className="row"
  />
</main>
`;
